src/Features/Core/Portable/ConvertLinq/ConvertForEachToLinqQuery/AbstractConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+
+namespace Microsoft.CodeAnalysis.ConvertLinq.ConvertForEachToLinqQuery
+{
+    internal abstract class AbstractConvertForEachToLinqQueryProvider<TForEachStatement, TStatement> : CodeRefactoringProvider
+        where TForEachStatement : TStatement
+        where TStatement : SyntaxNode
+    {
+        /// <summary>
+        /// Parses the forEachStatement until a child node cannot be converted into a query clause.
+        /// </summary>
+        /// <param name="forEachStatement"></param>
+        /// <returns>
+        /// 1) extended nodes that can be converted into clauses
+        /// 2) identifiers introduced in nodes that can be converted
+        /// 3) statements that cannot be converted into clauses
+        /// 4) trailing comments to be added to the end
+        /// </returns>
+        protected abstract ForEachInfo<TForEachStatement, TStatement> CreateForEachInfo(TForEachStatement forEachStatement);
```

Comment:
I think you use original tree/semantic model as you analyze and only update/transform code at the end. so might be convenient if you put semantic model (semantic info) in ForEachInfo as well? not just syntactic info? saying since I see you pass around foreach info and semantic model pair around.